### ISSUE TITLE
Encountered a loss convergence issue during reproduction.

### ISSUE THREAD

**Issue by helldog-star** [BryceZhuo/HybridNorm#2]
Hello! we are very interested in the novel normalization order of HybridNorm and are trying to reproduce it based on LlamaFactory：）

However, we found that in the FFN part, if the hidden state after normalization is used for the residual connection (as implemented in the paper), the model does not converge during training. 

But if the hidden state before normalization is used for the residual connection (standard pre-norm), this issue does not occur. 

Below is our main implementation of the model layer and attention. What could be the possible reasons for this? Thank you very much!

### Attention forward

        def forward(
            self,
            hidden_states: torch.Tensor,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            past_key_value: Optional[Cache] = None,
            output_attentions: bool = False,
            use_cache: bool = False,
            cache_position: Optional[torch.LongTensor] = None,
        ):
            ...
            bsz, q_len, _ = hidden_states.size()

            query_states = self.q_proj(hidden_states)
            key_states = self.k_proj(hidden_states)
            value_states = self.v_proj(hidden_states)

            query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
            key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
            value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)

            query_states = self.q_layernorm(query_states)
            key_states = self.k_layernorm(key_states)
            value_states = self.v_layernorm(value_states)

            cos, sin = self.rotary_emb(value_states, position_ids)
            query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
            ...

### Layer forward

        def forward(
            self,
            hidden_states: torch.Tensor,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            past_key_value: Optional[Cache] = None,
            output_attentions: Optional[bool] = False,
            use_cache: Optional[bool] = False,
            cache_position: Optional[torch.LongTensor] = None,
        ):
            residual = hidden_states

            # Self Attention
            hidden_states, self_attn_weights, present_key_value = self.self_attn(
                hidden_states=hidden_states,
                attention_mask=attention_mask,
                position_ids=position_ids,
                past_key_value=past_key_value,
                output_attentions=output_attentions,
                use_cache=use_cache,
                cache_position=cache_position,
            )
            hidden_states = residual + hidden_states
        
            # MLP
            hidden_states = self.post_attention_layernorm(hidden_states)
            residual = hidden_states

            hidden_states = self.mlp(hidden_states)
            hidden_states = residual + hidden_states
            ...

### Loss

![Image](https://github.com/user-attachments/assets/b7f8515c-70d4-453c-b287-b4279387f29a)

However, with a FFN pre-norm implementation, the loss decreases normally.

### Layer forward

        def forward(
            self,
            hidden_states: torch.Tensor,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            past_key_value: Optional[Cache] = None,
            output_attentions: Optional[bool] = False,
            use_cache: Optional[bool] = False,
            cache_position: Optional[torch.LongTensor] = None,
        ):
            residual = hidden_states

            # Self Attention
            hidden_states, self_attn_weights, present_key_value = self.self_attn(
                hidden_states=hidden_states,
                attention_mask=attention_mask,
                position_ids=position_ids,
                past_key_value=past_key_value,
                output_attentions=output_attentions,
                use_cache=use_cache,
                cache_position=cache_position,
            )
            hidden_states = residual + hidden_states
        
            # MLP
            residual = hidden_states
            hidden_states = self.post_attention_layernorm(hidden_states)
            
            hidden_states = self.mlp(hidden_states)
            hidden_states = residual + hidden_states
            ...

### Loss

![Image](https://github.com/user-attachments/assets/78d70b5a-6b44-46e4-b150-97b32e63d77e)

This is an interesting problem. Did you encounter similar issues during your experiments?

**Comment by BryceZhuo** [comment]
Hi！Thank you for reaching out—we’re pleased to hear that you’re exploring HybridNorm and HybridNorm$^*$ in your experiments.

In our experience with both OLMo2 and OLMoE, we did not observe convergence issues when employing HybridNorm or HybridNorm*. Based on the code snippets you provided, your implementation appears to be correct. Indeed, within Transformer architectures, Pre-Norm generally offers greater training stability than Post-Norm.

Given the limited information, we hypothesize that the convergence difficulties you’re encountering may be attributable to one or more of the following factors:

**1.	Initialization Method:** Appropriate parameter initialization is essential for stable training. In our experiments, we adopted the Megatron initialization, which yielded consistent and stable results. We encourage you to consult the “Ablation Studies → Initialization” section of our paper for further details. Suboptimal initialization may lead to gradient instability, especially in the early stages of training.

**2.	Number of Warmup Tokens:** Learning rate warmup is crucial for stabilizing the initial phase of training. Our setup utilized the default warmup token counts from OLMo2 and OLMoE, which were carefully tuned to ensure robust convergence across different normalization strategies. An insufficient warmup duration may lead the optimizer into unfavorable regions of the parameter space, particularly when normalization influences gradient scales or update dynamics.

We strongly recommend aligning your initialization and warmup configurations with those described in our paper and code， and implemented in the official OLMo2/OLMoE codebases. 

We hope this is helpful. Should the issue persist, please feel free to share additional details regarding your training setup (e.g., optimizer type, learning rate schedule, batch size, gradient clipping), and we would be glad to investigate further.

**Comment by helldog-star** [comment]
Thank you for your detailed response! I'll explore further based on these suggestions and update as soon as possible.

**Comment by helldog-star** [comment]
Hi, we conducted an experiment using the model configuration from HybridNorm and found that the loss converges normally. 

Through further analysis, we believe the issue lies with the number of layers—our model configuration has 24 layers. 

This is reasonable, as pre-norm tends to be more beneficial for training deeper models. :)
